### PR TITLE
use  peterbay/uvc-gadget

### PR DESCRIPTION
**Pull request by dgsiegel**
This gives us better performance, v4l2 controls and multiple resolutions. It also might be possible to merge that repo into showmewebcam

### CODE REVIEW

**Review by tuyenld — COMMENTED** [review]
Hi @dgsiegel,

Does this patch come along with your PR #26 ?
I tried to apply this patch with the current kernel.
A `kernel panic` happened when I open [webcamoid](https://webcamoid.github.io/) on Debian.
```
[  249.248761] configfs-gadget gadget: uvc_function_disable
[  249.326025] Unable to handle kernel paging request at virtual address 4ebcb7a8
[  249.336560] pgd = ddf5371b
[  249.342281] [4ebcb7a8] *pgd=00000000
[  249.348747] Internal error: Oops: 80000005 [#1] ARM
[  249.356487] Modules linked in: usb_f_acm u_serial usb_f_uvc bcm2835_codec(C) bcm2835_v4l2(C) v4l2_mem2mem bcm2835_mmal_vchiq(C) v4l2_common videobuf2_dma_6
[  249.399435] CPU: 0 PID: 0 Comm: swapper Tainted: G         C        4.19.113 #1
[  249.409944] Hardware name: BCM2835
[  249.416478] PC is at 0x4ebcb7a8
[  249.422737] LR is at __handle_irq_event_percpu+0x48/0x1d8
[  249.431264] pc : [<4ebcb7a8>]    lr : [<c0068484>]    psr: a0000193
[  249.440716] sp : c0a91df8  ip : c0a91db8  fp : c0a91e34
[  249.449070] r10: c0a91e38  r9 : 00000000  r8 : 00000000
[  249.457411] r7 : 00000021  r6 : cf031100  r5 : c0b1c790  r4 : cebcb840
[  249.467060] r3 : 4ebcb7a8  r2 : 00000000  r1 : 4ebcb7a8  r0 : 00000021
[  249.476799] Flags: NzCv  IRQs off  FIQs on  Mode SVC_32  ISA ARM  Segment user
[  249.487344] Control: 00c5387d  Table: 0dc94008  DAC: 00000055
[  249.496400] Process swapper (pid: 0, stack limit = 0x24cd8057)
[  249.505621] Stack: (0xc0a91df8 to 0xc0a92000)
[  249.513344] 1de0:                                                       00000001 0ca14a00
[  249.528397] 1e00: 0000003a cf031100 0000003a c0a93028 c0b1c790 cf031100 00000000 cf02e000
[  249.543858] 1e20: c0a90000 c0a68a30 c0a91e5c c0a91e38 c006864c c0068448 00000000 543c6d86
[  249.559554] 1e40: 00913bb0 cf031100 c0b1c790 00000001 c0a91e74 c0a91e60 c00686dc c0068620
[  249.575588] 1e60: cf031100 c0b1c790 c0a91e8c c0a91e78 c006bf44 c00686b0 00000021 c0b1c790
[  249.591641] 1e80: c0a91e9c c0a91e90 c00676dc c006be94 c0a91ec4 c0a91ea0 c0067dc0 c00676b8
[  249.607720] 1ea0: c0a91ee0 c0a93a0c ffffffff c0a91f14 c08c550c c0a90000 c0a91edc c0a91ec8
[  249.624074] 1ec0: c000a050 c0067d74 c0011584 40000013 c0a91f3c c0a91ee0 c000999c c000a020
[  249.640808] 1ee0: c0011580 00000000 40000093 40000093 ffffe000 00000000 c0a93090 c0b2149e
[  249.657847] 1f00: c08c550c 00000001 c0a68a30 c0a91f3c 00000000 c0a91f30 c00d659c c0011584
[  249.675266] 1f20: 40000013 ffffffff c0011580 00000000 c0a91f4c c0a91f40 c0756fc8 c0011560
[  249.692891] 1f40: c0a91f74 c0a91f50 c0050968 c0756f94 c0a9bb1c ffffffff c0a93028 c0b350c0
[  249.710779] 1f60: c0a93020 c0b350c0 c0a91f84 c0a91f78 c0050c60 c00508d4 c0a91f9c c0a91f88
[  249.728983] 1f80: c0751354 c0050c50 c0b35118 ffffffff c0a91ff4 c0a91fa0 c0a1cf10 c07512e4
[  249.747292] 1fa0: ffffffff ffffffff 00000000 c0a1c61c 00000000 cffffb00 00000000 c0a68a30
[  249.765592] 1fc0: 54387ef7 00000000 00000000 c0a1c330 00000055 00c0387d 00000c42 0ffe9f00
[  249.783896] 1fe0: 410fb767 00c5387d 00000000 c0a91ff8 00000000 c0a1cb28 00000000 00000000
[  249.802230] [<c0068484>] (__handle_irq_event_percpu) from [<c006864c>] (handle_irq_event_percpu+0x38/0x90)
[  249.822096] [<c006864c>] (handle_irq_event_percpu) from [<c00686dc>] (handle_irq_event+0x38/0x4c)
[  249.841193] [<c00686dc>] (handle_irq_event) from [<c006bf44>] (handle_level_irq+0xbc/0x12c)
[  249.859751] [<c006bf44>] (handle_level_irq) from [<c00676dc>] (generic_handle_irq+0x30/0x44)
[  249.878391] [<c00676dc>] (generic_handle_irq) from [<c0067dc0>] (__handle_domain_irq+0x58/0xb8)
[  249.897276] [<c0067dc0>] (__handle_domain_irq) from [<c000a050>] (bcm2835_handle_irq+0x3c/0x44)
[  249.916155] [<c000a050>] (bcm2835_handle_irq) from [<c000999c>] (__irq_svc+0x5c/0x7c)
[  249.929177] Exception stack(0xc0a91ee0 to 0xc0a91f28)
[  249.939335] 1ee0: c0011580 00000000 40000093 40000093 ffffe000 00000000 c0a93090 c0b2149e
[  249.957580] 1f00: c08c550c 00000001 c0a68a30 c0a91f3c 00000000 c0a91f30 c00d659c c0011584
[  249.975826] 1f20: 40000013 ffffffff
[  249.984266] [<c000999c>] (__irq_svc) from [<c0011584>] (arch_cpu_idle+0x30/0x40)
[  249.996702] [<c0011584>] (arch_cpu_idle) from [<c0756fc8>] (default_idle_call+0x40/0x48)
[  250.014525] [<c0756fc8>] (default_idle_call) from [<c0050968>] (do_idle+0xa0/0xf4)
[  250.027084] [<c0050968>] (do_idle) from [<c0050c60>] (cpu_startup_entry+0x1c/0x20)
[  250.039591] [<c0050c60>] (cpu_startup_entry) from [<c0751354>] (rest_init+0x7c/0x9c)
[  250.052202] [<c0751354>] (rest_init) from [<c0a1cf10>] (start_kernel+0x3f4/0x49c)
[  250.064474] Code: bad PC value
[  250.072188] ---[ end trace 1d0f00597e6511e5 ]---
[  250.081372] Kernel panic - not syncing: Fatal exception in interrupt
[  250.092311] ---[ end Kernel panic - not syncing: Fatal exception in interrupt ]---
```

**Comment by dgsiegel** [comment]
I've tried this patch both with the current kernel, as well as the newest one (#26), but never got a kernel panic. Did you build it yourself?

File: package/piwebcam/multi-gadget.sh [hunk 1]
```diff
@@ -1,58 +1,74 @@
...
-mkdir /sys/kernel/config/usb_gadget/pi4/functions/acm.usb0
-mkdir -p /sys/kernel/config/usb_gadget/pi4/functions/uvc.usb0/control/header/h
-ln -s /sys/kernel/config/usb_gadget/pi4/functions/uvc.usb0/control/header/h /sys/kernel/config/usb_gadget/pi4/functions/uvc.usb0/control/class/fs
-
-mkdir -p /sys/kernel/config/usb_gadget/pi4/functions/uvc.usb0/streaming/mjpeg/m/1080p
-cat <<EOF > /sys/kernel/config/usb_gadget/pi4/functions/uvc.usb0/streaming/mjpeg/m/1080p/dwFrameInterval
+
+CONFIG=/sys/kernel/config/usb_gadget/piwebcam
+mkdir -p $CONFIG
+cd $CONFIG
```
Comment (htruong) [inline package/piwebcam/multi-gadget.sh:5]: Those should be quoted and the whole script should be checked against [shellcheck](https://www.shellcheck.net/).

File: package/piwebcam/multi-gadget.sh [hunk 2]
```diff
@@ -1,58 +1,74 @@
...
+
+  echo $WIDTH > $FRAMEDIR/wWidth
+  echo $HEIGHT > $FRAMEDIR/wHeight
+  echo 333333 > $FRAMEDIR/dwDefaultFrameInterval
+  echo $(($WIDTH * $HEIGHT * 80)) > $FRAMEDIR/dwMinBitRate
+  echo $(($WIDTH * $HEIGHT * 160)) > $FRAMEDIR/dwMaxBitRate
+  echo $(($WIDTH * $HEIGHT * 2)) > $FRAMEDIR/dwMaxVideoFrameBufferSize
+  cat <<EOF > $FRAMEDIR/dwFrameInterval
 333333
+400000
```
Comment (htruong) [inline package/piwebcam/multi-gadget.sh:47]: Nice...

File: package/piwebcam/piwebcam.mk [hunk 3]
```diff
@@ -4,8 +4,8 @@
 #
 ################################################################################
 
-PIWEBCAM_VERSION = 29f6923730801980b48de444a362519728a62d7c
-PIWEBCAM_SITE = git://github.com/showmewebcam/uvc-gadget.git
+PIWEBCAM_VERSION = 4a025fc8911b0cb2bd49e6a3ca921b56006317f0
+PIWEBCAM_SITE = git://github.com/peterbay/uvc-gadget.git
```
Comment (htruong) [inline package/piwebcam/piwebcam.mk:8]: I wonder if we should cherry-pick @peterbay's commits and port it over here so we can have better control of what's going on in the long run.

**Comment by htruong** [comment]
> Does this patch come along with your PR #26 ?
> I tried to apply this patch with the current kernel.
> A `kernel panic` happened when I open [webcamoid](https://webcamoid.github.io/) on Debian.

@tuyenld 

I think you need both this PR's patch and PR#26 patch. I am a little bit of a github noob so I think to test this I would checkout the two PR and then create a new branch and cherry-pick both of the commits. We need both commits for your kernel to not panic, I think.

**Comment by tuyenld** [comment]
It happened once, but now it doesn't happen any more.

However, I cann't change the resolution. That is what I did:
```bash
# On host (Debian)
ldtuyen@pc:~$ v4l2-ctl --set-fmt-video=width=1920,height=1080
ldtuyen@pc:~$ 
ldtuyen@pc:~$ v4l2-ctl --get-fmt-video
Format Video Capture:
	Width/Height      : 1920/1080
	Pixel Format      : 'MJPG' (Motion-JPEG)
	Field             : None
	Bytes per Line    : 0
	Size Image        : 2880000
	Colorspace        : sRGB
	Transfer Function : Default (maps to sRGB)
	YCbCr/HSV Encoding: Default (maps to ITU-R 601)
	Quantization      : Default (maps to Full Range)
	Flags             : 
ldtuyen@pc:~$ 

# On device (Pi)
# v4l2-ctl --get-fmt-video
Format Video Capture:
        Width/Height      : 1600/1200
        Pixel Format      : 'MJPG' (Motion-JPEG)
        Field             : None
        Bytes per Line    : 0
        Size Image        : 1920000
        Colorspace        : SMPTE 170M
        Transfer Function : Default (maps to Rec. 709)
        YCbCr/HSV Encoding: Default (maps to ITU-R 601)
        Quantization      : Default (maps to Full Range)
        Flags             : 
# 
```
It is alway 1600x1200 dispite of my setting on Debian.

Currently Pi Cam is running 1920x1080, but I can't find the default resolution in the peterbay's code.
Do you know what is it?

**Comment by peterbay** [comment]
Hi,
my code doesn't use settings defined in uvc-gagdet. All available resolutions are obtained from configfs which are defined in multi-gadget.sh. Host computer obtain available resolutions from uvc-driver which reads settings from configfs. Then host computer send probe requests and commit request for choosing resolution. 
If uvc-gagdet is run manually then you can see all steps.

--- obtaining informations from configfs ---
```
CONFIGFS: Initial path: /sys/kernel/config/usb_gadget
CONFIGFS: UVC: format: 1, frame: 10, resolution: 1600x1200, bitrate: [153600000, 307200000]
CONFIGFS: UVC: format: 1, frame: 9, resolution: 1600x900, bitrate: [115200000, 230400000]
CONFIGFS: UVC: format: 1, frame: 8, resolution: 1536x864, bitrate: [106168320, 212336640]
CONFIGFS: UVC: format: 1, frame: 7, resolution: 1920x1080, bitrate: [165888000, 331776000]
CONFIGFS: UVC: format: 1, frame: 6, resolution: 1280x960, bitrate: [98304000, 196608000]
CONFIGFS: UVC: format: 1, frame: 5, resolution: 1280x720, bitrate: [73728000, 147456000]
CONFIGFS: UVC: format: 1, frame: 4, resolution: 1024x768, bitrate: [62914560, 125829120]
CONFIGFS: UVC: format: 1, frame: 3, resolution: 800x600, bitrate: [38400000, 76800000]
CONFIGFS: UVC: format: 1, frame: 2, resolution: 640x480, bitrate: [24576000, 49152000]
CONFIGFS: UVC: format: 1, frame: 1, resolution: 640x360, bitrate: [18432000, 36864000]
```


--- getting and setting of resolution ---
```
UVC: Streaming request CS: PROBE, REQ: GET_CUR
UVC: Streaming request CS: PROBE, REQ: SET_CUR
UVC: Control PROBE, length: 26
UVC: Streaming control: action: SET, format: 1, frame: 7
FRAME: format: 1, frame: 7, resolution: 1920x1080, bitrate: [165888000, 331776000]
DUMP: uvc_streaming_control: format: 1, frame: 7, frame interval: 333333
UVC: Streaming request CS: PROBE, REQ: GET_CUR
UVC: Streaming request CS: PROBE, REQ: GET_MAX
UVC: Streaming control: action: GET MAX
FRAME: format: 1, frame: 10, resolution: 1600x1200, bitrate: [153600000, 307200000]
DUMP: uvc_streaming_control: format: 1, frame: 10, frame interval: 333333
UVC: Streaming request CS: PROBE, REQ: GET_MIN
UVC: Streaming control: action: GET MIN
FRAME: format: 1, frame: 1, resolution: 640x360, bitrate: [18432000, 36864000]
DUMP: uvc_streaming_control: format: 1, frame: 1, frame interval: 333333
UVC: Streaming request CS: PROBE, REQ: SET_CUR
UVC: Control PROBE, length: 26
UVC: Streaming control: action: SET, format: 1, frame: 7
FRAME: format: 1, frame: 7, resolution: 1920x1080, bitrate: [165888000, 331776000]
DUMP: uvc_streaming_control: format: 1, frame: 7, frame interval: 333333
UVC: Streaming request CS: PROBE, REQ: GET_CUR
UVC: Streaming request CS: COMMIT, REQ: SET_CUR
UVC: Control COMMIT, length: 26
UVC: Streaming control: action: SET, format: 1, frame: 7
FRAME: format: 1, frame: 7, resolution: 1920x1080, bitrate: [165888000, 331776000]
DUMP: uvc_streaming_control: format: 1, frame: 7, frame interval: 333333
```

File: package/piwebcam/piwebcam.mk (hunk 3 above)
Comment (dgsiegel) [reply package/piwebcam/piwebcam.mk:8]: It might make sense to combine these two repositories into one?

**Comment by dgsiegel** [comment]
More specifically you need to patch your kernel, or at least the uvc module. This is what @peterbay told me:

> If you need resolution changing then f_uvc.c file must be changed
> 
> https://github.com/raspberrypi/linux/blob/rpi-5.4.y/drivers/usb/gadget/function/f_uvc.c#L817
> 
> !! must be compiled with a kernel !!
> This two lines are configured only for brightness
> pd->bmControls[0]		= 1;
> pd->bmControls[1]		= 0;
> and need to be set to all available controls
> pd->bmControls[0]		= 255;
> pd->bmControls[1]		= 255;

I didn't try it myself yet, but I know @peterbay has and he could control all settings.

**Comment by htruong** [comment]
@peterbay I just saw that your repo is very active. We will move our pointer to your repo then. Thanks for chiming in!

File: package/piwebcam/piwebcam.mk (hunk 3 above)
Comment (htruong) [reply package/piwebcam/piwebcam.mk:8]: Yeah let's direct it to @peterbay's repo.

File: package/piwebcam/multi-gadget.sh (hunk 1 above)
Comment (dgsiegel) [reply package/piwebcam/multi-gadget.sh]: I've updated the shell script. On a second thought I think this should be part of @peterbay's uvc-widget repo, as we don't configure anything related to showmewebcam there.

**Comment by htruong** [comment]
I have built the raspberry pi 0w from scratch with 4 of your patches applied and now I cannot boot successfully. My pi hangs after outputting something about `brcmfmac`...  Trying to figure out what went wrong.

**Comment by dgsiegel** [comment]
Sorry, I've somehow messed up my branch and inserted some syntax errors in the multi-gadget.sh file, hence you couldn't boot. Would you mind checking out this branch again (I've used push -f) and retry?

**Comment by htruong** [comment]
Thanks for the change and the update to the PR!

**Pull request merged**